streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/ClockService.scala

```diff
@@ -36,71 +38,84 @@ import org.slf4j.Logger
 
 import scala.concurrent.Future
 import scala.concurrent.duration.FiniteDuration
-import scala.collection.JavaConversions._
 
 /**
  * The clockService will maintain a global view of message timestamp in the application
  */
 class ClockService(dag : DAG, store: AppDataStore) extends Actor with Stash {
-
   private val LOG: Logger = LogUtil.getLogger(getClass)
 
   import context.dispatcher
 
-  private var startClock: Long = 0
-  
-  private val processorClockLookup = new util.HashMap[ProcessorId, ProcessorClock]()
-
-  private var reportScheduler : Cancellable = null
+  private val healthChecker = new HealthChecker
+  private var healthCheckScheduler : Cancellable = null
   private var snapshotScheduler : Cancellable = null
-  private var stallingTasks = List.empty[StallingTask]
 
-  var processorIdToLevel = Map.empty[ProcessorId, Int]
-  var levelMinClock = Array.empty[TimeStamp]
-  
   override def receive = null
 
   override def preStart() : Unit = {
     LOG.info("Initializing Clock service, get snapshotted StartClock ....")
-
     store.get(START_CLOCK).asInstanceOf[Future[TimeStamp]].map { clock =>
       val startClock = Option(clock).getOrElse(0L)
       self ! StartClock(startClock)
-      LOG.info(s"Start Clock Retrived, starting ClockService, startClock: $startClock")
+      LOG.info(s"Start Clock Retrieved, starting ClockService, startClock: $startClock")
     }
 
     context.become(waitForStartClock)
   }
 
   override def postStop() : Unit = {
-    Option(reportScheduler).map(_.cancel)
+    Option(healthCheckScheduler).map(_.cancel)
     Option(snapshotScheduler).map(_.cancel)
   }
 
-  private def initializeDagWithStartClock(startClock: TimeStamp) = {
-    this.startClock = startClock
-    dag.processors.foreach {
-      processorIdWithDescription =>
-        val (processorId, description) = processorIdWithDescription
-        val taskClocks = new Array[TimeStamp](description.parallelism).map(_ => startClock)
-        val processorClock = new ProcessorClock(taskClocks)
-        processorClockLookup.put(processorId, processorClock)
+  private var clocks = Map.empty[ProcessorId, ProcessorClock]
+  private var upstreamClocks = Map.empty[ProcessorId, Array[ProcessorClock]]
+  private var sinkClocks = Array.empty[ProcessorClock]
```

Comment:
IMO. better use an immutable data structure for `sinkClocks` since it references processor clocks which should not be reset.